### PR TITLE
G4 - Decorator Library

### PR DESCRIPTION
**Pull request by seg2151**
Jason
Sebastian

### CODE REVIEW

File: decorators_library/decorators.py [hunk 1]
```diff
@@ -1 +1,89 @@
-# implement your decorators here.
+import signal
+from .exceptions import TimeoutError
+import logging
+import time
+
+def timeout_occured(signum, frame):
```
Comment (janiceshiu) [inline decorators_library/decorators.py:6]: interesting idea and implementation! will look at the documentation for `signal` when I have more time

File: decorators_library/decorators.py [hunk 2]
```diff
@@ -1 +1,89 @@
...
+    return logging_decorator
+
+class count_calls():
+    counters_dict = {}
+    
+    def __init__(self, func):
+        self.func = func
+        self.counters_dict[func.__name__] = 0
+        
+    def __call__(self, *args, **kwargs):
```
Comment (janiceshiu) [inline decorators_library/decorators.py:43]: we also used `**kwargs` iirc are still not sure whether it is necessary. wonder whether it's necessary, hm hm

File: decorators_library/decorators.py [hunk 3]
```diff
@@ -1 +1,89 @@
...
+    @classmethod
+    def reset_counters(cls):
+        cls.counters_dict = {}
+        
+    def counter(self):
+        return self.counters_dict.get(self.func.__name__, 0)
+        
+def memoized(func):
+    def memo_function(*args, **kwargs):
+        if not hasattr(memo_function, "cache"):
```
Comment (janiceshiu) [inline decorators_library/decorators.py:61]: nice - I like this

File: tests/test_decorators.py [hunk 4]
```diff
@@ -100,3 +101,38 @@ def add(a, b):
         self.assertEqual(add.cache, {(1, 2): 3, (2, 3): 5})
         self.assertEqual(add(3, 4), 7)
         self.assertEqual(add.cache, {(1, 2): 3, (2, 3): 5, (3, 4): 7})
+
+    def test_stopwatch(self):
+        @stopwatch
+        def add(a, b):
+            return a + b
+        add(1,3)
```
Comment (janiceshiu) [inline tests/test_decorators.py:109]: readability. space after comma.

**Review by janiceshiu — COMMENTED** [review]
nice job implementing the tested ones and adding your own! learnt a lot and will google up more later. congrats on doing it so fast as well.

File: decorators_library/decorators.py (hunk 2 above)
Comment (jangeador) [reply decorators_library/decorators.py:43]: When you write a decorator it's usually for general use. `**kwargs` might not be needed for the current function that you are decorating, but it is nice to know that your decorator would handle keyword arguments if another function requires it. Always go with _args and *_kwargs if possible to give your decorator more usability.

Comment (janiceshiu) [reply decorators_library/decorators.py:43]: gotcha. thanks!

**Pull request closed without merging**